### DISCUSSION TITLE
Definition recursion and struct disjunction

### DISCUSSION THREAD

**Discussion by b4nst** [cue-lang/cue#1186] [Q&A]
So first sorry for the title, I does not really know where the important part is in the following question

So I have this structure:

```txtar
-- invalid.yaml --
components:
  one:
    components:
      two:
        name: foo
-- main.cue --
#Node: {
	id?:   string
	name?: string
	components: [string]: #Leaf | #Node
}

#Leaf: {
	id: string
}
-- valid.yaml --
components:
  one:
    components:
      two:
        id: foo
```

I was expecting that `cue vet invalid.yaml main.cue -d '#Node'` would fail and that `cue vet valid.yaml main.cue -d '#Node'` would succeed when it's actually exactly the opposite.

Any idea what I'm missing there?

**Comment by seh** [comment]
`#Node.components` is not a required field, per #822, so "components.one.components. two" satisfies `#Node`, which tolerates a "name" field.

Unifying with `#Node` shows that [the "components" field winds up in there](https://tip.cuelang.org/play/?id=sJSPrvV5B1m#cue@export@cue), but is not required to be there from _cue vet_'s perspective, at least for today.

**Reply by b4nst** [reply]
Oh wow, for some reason I thought `#Node.components` was required but now it makes total sense. Thanks for the enlightening!

**Reply by seh** [reply]
That's how I've read it too. You can see by comments starting at https://github.com/cue-lang/cue/issues/822#issuecomment-873390104 that I shared your interpretation. I'm coming around to believe that #822 is necessary, but the present state of CUE is confusing. It seems to me that there are two different _lenses_ or _perspectives_ through which one can read the same CUE code—say, _export_ versus _vet_, or generation versus validation—by which the same code means something different in each context.

**Reply by myitcv** [reply]
Yep, the discussion in #822 is relevant here. 

But just to note, do you even need `#Leaf` here? Is a `#Leaf` not simply a `#Node` without a `components` field defined?

The builtins discussed in https://github.com/cue-lang/cue/issues/943 will also be of interest (they are strongly linked to #822)

To also note that the disjunction `#Leaf | #Node` is problematic because of the reasons described in https://github.com/cue-lang/cue/discussions/1163

**Reply by b4nst** [reply]
`#Leaf` is not exactly a `#Node` without a `components` field. I wanted to describe the constraint of having an id set on leafs, while it's optional on a node.

**Reply by myitcv** [reply]
Ah I see, ok. You definitely either need #822 or some other mechanism to disambiguate `#Node` and `#Leaf`
